Use bind address for consul checks (solves #1574)

evan2645 opened the pull request:
Suggested solution for https://github.com/hashicorp/nomad/issues/1574

dadgar commented on command/agent/agent.go:
```diff
@@ -418,6 +418,16 @@ func (a *Agent) setupServer() error {
 				},
 			},
 		}
+		// Resolve ServiceCheck addresses
```
Can you pull these into three methods, `getHTTPAddr(preferBind bool)`, `getRPCAddr(preferBind bool)` and `getSerfAddr(preferBind bool)` and use those methods for the service and check. 

https://github.com/hashicorp/nomad/blob/master/command/agent/agent.go#L175-L188

Then add a config option to the `Consul` config `checks_use_advertise bool`. This will default to false, but if people want the checks to use the advertise they can set it to true.

Please add tests and documentation to the website

dadgar commented:
@evan2645 Do you want to close this?

In 0.5.1 I think we are going to take a fresh look at how we do network binding/advertising so we can tackle this then?

evan2645 commented:
@dadgar I have been meaning to clean this up but have not yet had the time. I am still willing to do this work if you think it is useful, however if it will be refactored/re-written soon, maybe not.

We are currently running a fork with this patch applied. How far away do you think 0.5.1 is? Getting back on upstream is something we'd like to do.

dadgar commented:
Okay I was just not sure if it was abandoned :)

So the comment I suggested still applies with the refactor so it is worth while 👍

evan2645 commented:
@dadgar I pushed the changes requested.

There are actually three modes we're looking for (real bind, advertise, and preferBind)... so I added a fourth method to account for the preferBind state. I simplified all the code setting bind and advertise config in all the places i could find it. added tests and docs. not sure if this is what you had in mind or not... lmk if it does not suffice

evan2645 commented:
fwiw, i'm pretty sure that failing test is unrelated to these changes

evan2645 commented:
would like to see this get merged sometime soon, we're still running a fork with it

i'm more than happy to commit additional changes or open a new pr if this is not the direction you had in mind... just let me know what i can do to get it solved

dadgar commented:
@evan2645 I am going to take this PR over as we are close to cutting an RC. Sorry for the delay it has been busy around here :)
